coalib/core/Core.py

```diff
@@ -327,34 +150,177 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
     return dependency_tracker, bears
 
 
-def run(bears, result_callback):
+class Session:
     """
-    Runs a coala session.
+    Maintains Session for a coala execution. For each session, there are set
+    of bears along with callback function, which is run when results are
+    available. On running a session, first its dependencies are initalized and
+    all the bears are scheduled to run in appropriate order.
 
-    :param bears:
-        The bear instances to run.
-    :param result_callback:
-        A callback function which is called when results are available. Must
-        have following signature::
+    Ex: If BearA requires BearB as dependencies, then, on running BearA,
+        first BearB will be executed, and after that, BearA will start running.
+    """
+
+    def __init__(self, bears, result_callback):
+        """
+        :param bears:
+            The bear instances to run.
+        :param result_callback:
+            A callback function which is called when results are available.
+            Must have following signature::
 
             def result_callback(result):
                 pass
-    """
-    # FIXME Allow to pass different executors nicely, for example to execute
-    # FIXME   coala with less cores, or to schedule jobs on distributed systems
-    # FIXME   (for example Mesos).
-
-    # Set up event loop and executor.
-    event_loop = asyncio.SelectorEventLoop()
-    executor = concurrent.futures.ProcessPoolExecutor()
-
-    # Initialize dependency tracking.
-    dependency_tracker, bears_to_schedule = initialize_dependencies(bears)
-
-    # Let's go.
-    schedule_bears(bears_to_schedule, result_callback, dependency_tracker,
-                   event_loop, {}, executor)
-    try:
-        event_loop.run_forever()
-    finally:
-        event_loop.close()
+        """
+        self.bears = bears
+        self.result_callback = result_callback
+
+    def run(self):
+        """
+        Runs a coala session.
+        """
+        # FIXME Allow to pass different executors nicely, for example to execute
+        # FIXME   coala with less cores, or to schedule jobs on distributed
+        # FIXME   systems (for example Mesos).
+
+        # Set up event loop and executor.
+        self.event_loop = asyncio.SelectorEventLoop()
+        self.executor = concurrent.futures.ProcessPoolExecutor()
+        self.running_tasks = {}
+
+        # Initialize dependency tracking.
+        self.dependency_tracker, bears_to_schedule = initialize_dependencies(
+            self.bears)
```

Comment:
Good point. What about `executor' ? There comments in the code indicating that we'll allow custom executor in future. In that case, if we move `self.executor = concurrent.futures.ProcessPoolExecutor()` in constructor, then it will add additional cost if custom executor is used. I don't know how much expensive this call is.
If it is not much expensive, we can do it constructor. Otherwise, we can do have an param in `run` function called `executor` and code in the function should look something like `self.executor = executor if executor is not None else concurrent.futures.ProcessPoolExecutor()`.
But that change is not in scope of this PR.
What do you think ?